lib/Serialization/Serialization.cpp

```diff
@@ -2918,7 +2918,8 @@ void Serializer::writeDecl(const Decl *D) {
     auto genericParam = cast<GenericTypeParamDecl>(D);
     verifyAttrSerializable(genericParam);
 
-    auto contextID = addDeclContextRef(genericParam->getDeclContext());
+    auto contextID = addDeclContextRef(genericParam->getDeclContext()
+                                         ->getParent());
```

Comment:
I don't understand this change. Why is it okay to change the context of the generic parameter?

Reply:
We deserialize (and parse) generic parameter decls before the parent type, then change their context when we build the type. I agree it’s iffy but a) we’ve always done this b) nothing should use the context of a generic parameter.

We used to serialize the “right” context, but ignore it when deserializing, to avoid a cycle between the generic parameter decl and its owner decl. This change makes it more explicit by serializing the “wrong” context, but it shouldn’t change behavior.

Reply:
At parse time, generic parameter decls serve as name lookup results before the signature has been built; a lookup of “T” has to return _something_ that’s a decl.

After deserialization though, they don’t serve much purpose except for storing the sugared name. If we had a way to store the name directly inside the deserialized generic parameter type, we wouldn’t need to serialize generic parameter decls (or lists) at all.

Certainly nothing should be emitting xrefs to generic parameter decls, or looking at their context, etc. Such code would fall over with canonical types anyway.

Reply:
If they don't need a context at all, can we just not serialize it?

Reply:
You have to set it to _something_ when you deserialize the decl though, no? Maybe just the module would work...

Reply:
Yeah, if it can't be `nullptr` when it's first created, the module is as good as anything.